### DISCUSSION TITLE
How to choose different styles according to the user's device when using SSG or SSR

### DISCUSSION THREAD

**Discussion by yakiler** [vercel/next.js#44368] [Help] [answered]
I have a page expecting it to be styled differently on PC or mobile and I wish to take advantage of SSG. Is it feasible?
     Added: I need the same page to use different components depending on the device (they may be styled differently)

**Comment by GabenGar** [comment]
Depends on your stack. You can assign [CSS custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/--*) and overwrite their values within `@media` queries depending on the screen size.

**Reply by yakiler** [reply]
Thank you very much. I also need to select different components (userAgent gets device) according to the device (so the style will not be the same)

**Reply by yakiler** [reply]
At the same time, I don’t want to lose the advantages of static pages. Can NextJs collect User Agent to generate static pages after user requests?

**Reply by GabenGar** [reply]
No, because it has no idea of the agent at build time, since there is no actual browser request happening.
You either get imperative `userAgent()` agent calls or you get static pages, can't have both.

**Comment by icyJoseph** [comment, marked as answer]
I think the only way to try and do this is by pre-rendering every possible version of a page. Of course assuming there's a well-defined and finite set of versions you want to generate.

```ts
// middleware.ts
import { NextRequest, NextResponse, userAgent } from "next/server";

export async function middleware(req: NextRequest) {
  const { pathname, href } = req.nextUrl;

  const ua = userAgent(req);

  // Here I am just throwing some silly logic, you'd need to derive the type of device yourself
  if (pathname === "/about" && ua.browser.name === "Chrome") {
    req.nextUrl.pathname = "/_screen/narrow/about";
    return NextResponse.rewrite(req.nextUrl);
  } else if (pathname === "/about") {
    req.nextUrl.pathname = "/_screen/desktop/about";
    return NextResponse.rewrite(req.nextUrl);
  }
  return NextResponse.next();
}
```

And then you could have a page like:

```tsx
// pages/_screen/[size]/about.tsx
import { GetStaticPaths, GetStaticPropsContext, type NextPage } from "next";
import { useRouter } from "next/router";

const About: NextPage = () => {
  const { query } = useRouter();

  const { size } = query;
  return <div>This is the about page: {size}</div>;
};

export const getStaticPaths: GetStaticPaths<{ size: string }> = () => {
  return {
    fallback: false,
    paths: [{ params: { size: "desktop" } }, { params: { size: "narrow" } }],
  };
};

export const getStaticProps = (ctx: GetStaticPropsContext) => {
  const params = ctx.params;

  const size = params?.size;

  if (!size) throw new Error("Size is missing as param");

  return { props: {} }; // you could send the size as props too
};

export default About;
```

So now when I go to this page on Chrome I see the narrow version, and when I use Safari I see the desktop version. Best of all, is that these are pre-rendered.

```
Route (pages)                              Size     First Load JS
┌ ○ /                                      1.96 kB          78 kB
├   /_app                                  0 B              74 kB
├ ● /_screen/[size]/about                  359 B          74.4 kB
├   ├ /_screen/desktop/about
├   └ /_screen/narrow/about
├ ○ /404                                   181 B          74.2 kB
└ λ /api/hello                             0 B              74 kB


ƒ Middleware                               28.9 kB

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
   (ISR)     incremental static regeneration (uses revalidate in getStaticProps)

```

I'll reserve my opinion on the maintainability of this approach, but since it looks like you really want to just get it done there it goes.

**Reply by yakiler** [reply]
It looks like it does exactly what I need, which is great. Thanks.
